### ISSUE TITLE
Meta Tags Are Missing

### ISSUE THREAD

**Issue by instantflorian** [WBCE/WBCE_CMS#402]
In the frontend and in the backend the <title> tags is missing, and also all other metatags which were usually generated (description, keywords, charset)

**Comment by mrbaseman** [comment]
when doFilter is called a second time an already inserted title, keywords or description are replaced. 
If the corresponding queue is cleared before, the meta entries are lost. the above fix should restore them making use of the fact that the queue still holds the entries that were used before.

**Comment by WebDesignWorx** [comment]
Hello @mrbaseman , didn't see your fix in the morning hours and created another one.
Maybe you like to have a look at it.

**Comment by instantflorian** [comment]
Sorry for double work. The problem is that I do not know how to adress both of you at the same time. mrbaseman prefers Github, but much of communication and testing is done in the forum.

Unfortunately, both of the solutions which were provided do not work properly. mrbaseman's commit fixes only the issue in the backend but not in the frontend.

Stefek's PR fixes the issue on the surface, but then again the head data is generated twice.

**Comment by WebDesignWorx** [comment]
> Stefek's PR fixes the issue on the surface, but then again the head data is generated twice.
Can't reproduce this behavior.
I don't mind using GitHub for communication more extensively, however, reading emails is not the first action in the morning for me, thus I missed @mrbaseman's entry.

However, we need to handle it in a way that diFilter() is called only once throughout the system. Doing otherwise will bring about an array of other issues and broken functionality.

**Comment by mrbaseman** [comment]
Well, the AutoPlaceholder filter contains its own Regex Patterns (the same ones as the insert class). We could use that for inserting the place holders and wait with the insert class until the end.

**Comment by WebDesignWorx** [comment]
I don't see how it could help resolving the issue @instantflorian just mentioned (which I can't even reproduce), but sure, we can give it a try.

**Comment by instantflorian** [comment]
Solved with https://github.com/WBCE/WBCE_CMS/commit/5413bf5d4ee693c566655d713cc8ff101ea71e88

**Comment by WebDesignWorx** [comment]
BTW. Since class Insert is integral part of the core and is always present since WBCE 1.4.0 we could just skip [this entire code (emphasized) ](https://github.com/WBCE/WBCE_CMS/blob/76c1f46cf679ff181ba694a8eb2c5d5fc43f17bb/wbce/modules/mod_opf_auto_placeholder/filter.php#L40-L106)

**Comment by mrbaseman** [comment]
It seems you have solved this already this morning, and the current solution actually was what I would have proposed now, too. Sorry, I should have had a closer look at the code of the Auto Placeholder Filter earlier.
 @instantflorian what about PM in the forum? It could well be a short notice "have a look at thread https://xy". My main problem is that the forum does not reliably inform me about new posts. I'm following the rss feed and I receive the email notifications, but the notifications are sent out only once for each thread until I have read that particular message again online (after I have seen it in the email already). I have to log in and then the font changes to something unreadable, microscopic on the mobile. I believe that I'm informed about every PM, regardless if I have read the previous one (but I'm not really sure about that). 
@WebDesignWorx yes, I'll remove that code. I have backported the filter modules from the dev-2.x branch before the insert class was available in the current 1.4 development branch. Once the insert class was there I have left that code as a fallback but the chances that the filter modules are used somewhere where the insert class is not available are quite low. I'll clean up the filter modules.

**Comment by mrbaseman** [comment]
I have cleaned up the filter modules and updated the requirements in their precheck.php with https://github.com/WBCE/WBCE_CMS/commit/7e659cfbfcdf918c489cc5ba309f267deedc50c8

**Comment by WebDesignWorx** [comment]
No problem at all, @mrbaseman .
I'm very satisfied with all the progress we have all done as a team.
Now OpF Dashboard and the Insert class working hand in hand, flawlessly.

As to your forum notifications:
you could bookmark this link: https://forum.wbce.org/search.php?action=show_recent
It will show you all recent, unread posts.

**Comment by instantflorian** [comment]
@mrbaseman There are some things to say about the forum.

- First of all, as you are an early member, I see in your profile that you still use the old "Air" theme, which is indeed not responsive. Please switch to Aira2, which works on mobile displays too.
- Second, the threads in the forum about the ongoing development are in a restricted area which is only accessable for developers. If you are not logged in, therefore these threads neither appear in the RSS feed, nor in the Recent posts. This is not a bug but wanted behavior.
- Third, about notifications. Any forum software I know behaves this way, that you only get a message about the first post in a topic or thread you subscribed for, and not every further answer. Look at Simple Machines Forum, MyBB, phpBB, they all behave the same. So it's not a question of realibility of the forum software.

I am strictly against heavily use of the PM system of the WBCE forum. The PM system there is quite rudimentary, you can neither search nor save PMs to subfolders, so that you have to delete PMs frequently, which is really a pity when trying to recall how decisions where made lateron. Remember our quarrel about the abandoned 1.2 version? It is really important for me as project owner to have the possibility to proove that and how decisions where made.

So which options to we have?
1) I think the easiest way will be using the Github issue system, as notifications are sent out for every post and it's quite easy to refer to lines of code and so on. Disadvantage: All communication is public.
2) Forum: Development can be discussed without public view, which makes it easier to refer to non-public testing environments and of course is necessary if security related things are discussed. Disadvantage: As stated before, one has to be logged-in and messages are only sent on the first not-readed post in a subscribed topic.
3) Forum PM: Just No.
4) Ticket / bugtracker system: If you would say it would be helpful, I can set up an extra bugtracker system like Mantis or Bugzilla, but to be honest, I'm not very keen on this, although I'm willing to do everything which might ensure the further development of WBCE.

BR
-Florian.

**Comment by mrbaseman** [comment]
Hi @instantflorian 

> First of all, as you are an early member, I see in your profile that you still use the old "Air" theme, which is indeed not responsive. Please switch to Aira2, which works on mobile displays too.

thanks for the hint. I have changed that and it helps a lot

> Second, the threads in the forum about the ongoing development are in a restricted area which is only accessable for developers. If you are not logged in, therefore these threads neither appear in the RSS feed, nor in the Recent posts. This is not a bug but wanted behavior.

I know that. Therefore, the notifications are important. Otherwise I have to log into the forum after each email that I receive, quickly check that thread so that I still get informed about new posts. The problem with that procedure is that when I log out then, the forum marks all other topics as not new anymore. So, on each login I have to check all the threads which contain new posts.

> Third, about notifications. Any forum software I know behaves this way, that you only get a message about the first post in a topic or thread you subscribed for, and not every further answer. Look at Simple Machines Forum, MyBB, phpBB, they all behave the same. So it's not a question of realibility of the forum software.

SMF gives me the choice if I want to be notified about the first post of a thread or if I want to be informed each time. I still have to subscribe to new topics, but at least about the ones I'm on I'll get all posts by mail.

> I am strictly against heavily use of the PM system of the WBCE forum....

I didn't mean to use PM for the full discussion, but if there is something important that I shall know, dropping me the link to the post via PM would perhaps be an option.

> So which options to we have?
> 1. I think the easiest way will be using the Github issue system,...

I agree

> 2. Forum: Development can be discussed without public view,..

Yes, that's an advantage, too. And well, before pushing a commit on github or before investigating an issue I can double-check if there is a discussion relevant to that issue in the forum. It's just an extra step that I miss sometimes when I see a notification about a new issue on github. 

> 3. Forum PM: Just No.

agreed. See above

> 4. Ticket / bugtracker system: 

I don't think that another place for discussions makes the situation better. It makes only sense if we agree that all discussions should take place there. But given the fact that the forum is an established platform for discussions and the code is hosted on github, I believe moving the discussions to a third place would not be an unrealistic scenario. It has to do with habits. People tend to use what they personally prefer and what they are already using. So, there is some barrier when a topic is discussed in the forum and one would have to move the discussion each time to the bug tracker in order to discuss a possible solution.

So, let's stay with the forum and github even if our personal habits how to organize the work sometimes may lead to the situation where two of us notice too late that we work in parallel on the same topic. We can try to avoid it, but even if we would use only one platform, it could still happen that someone looks at the code, does some debugging offline and then notices that someone else did the same at the same time. Trying to avoid that would end in a very formalized workflow which in turn would also cause a lot of overhead... and apart from that it would kill a lot of enthusiasm and fun that we have when working more spontaneously.

best regards
Martin

**Comment by WebDesignWorx** [comment]
Agree on all accounts.
And by the way, it wasn't such a big deal. My solution consisted of 3 lines of code if I remember right. :-) 

Sometimes I think it would be great to have some kind of overview of tasks and subtasks.
We could use something like Todoist (which I use for my professional work) or Xmind for creating a better overview of all open and future subprojects and ideas. There are many similar tools, free and paid services aswell.
However, if it does create too much administrational overhead, let's stick to what we have: GitHub and Forum.

Kind regards, 
Christian
